Studies/source/EfficiencyTriggerStudy.cxx

```diff
@@ -0,0 +1,351 @@
+public:
+    using Channel = analysis::Channel;
+
+    struct SampleDesc {
+        std::vector<double> points_list;
+        std::string file_name;
+        std::string x_title;
+        analysis::EllipseParameters massWindowParams;
+    };
+
+    struct TriggerDesc {
+
+        std::vector<std::string> trigger;
+        std::string title;
+        root_ext::Color color;
+        analysis::TauIdDiscriminator tauId;
+        DiscriminatorWP tauIdWP;
+
+       };
+
+    EfficiencyStudy(const Arguments& _args) :
+      args(_args), canvas("","", 600, 600)
+    {
+      gStyle->SetOptStat(0);
+      canvas.SetGrid();
+      canvas.Print((args.output_file() + ".pdf[").c_str());
+      canvas.Draw();
+
+      sample_desc = LoadConfig(trigger_descs);
+    }
+
+    static TEfficiency CreateEfficiency(const TH1D& passed, const TH1& total)
+    {
+        constexpr static double oneSigma = 0.682689492137;
+
+        if(!TEfficiency::CheckConsistency(passed, total))
+            throw exception("passed TEfficiency objects do not have consistent bin contents");
+
+        TEfficiency eff (passed, total);
+        eff.SetConfidenceLevel(oneSigma);
+        eff.SetStatisticOption(TEfficiency::kFCP);
+
+        return eff;
+    }
+
+    static std::shared_ptr<TGraphAsymmErrors> CreateTGraph(const TEfficiency& eff,
+                                                           const std::vector<double>& mass_list)
+    {
+        std::vector<double> eff_value(mass_list.size());
+        std::vector<double> eff_errorUp(mass_list.size());
+        std::vector<double> eff_errorLow(mass_list.size());
+        std::vector<double> eff_error_x(mass_list.size(), 0.);
+
+        for(size_t n = 0; n < mass_list.size(); n++){
+             eff_value.at(n) = eff.GetEfficiency(static_cast<int>(n+1));
+             eff_errorUp.at(n) = eff.GetEfficiencyErrorUp(static_cast<int>(n+1));
+             eff_errorLow.at(n) = eff.GetEfficiencyErrorLow(static_cast<int>(n+1));
+         }
+
+        return std::make_shared<TGraphAsymmErrors>(static_cast<int>(mass_list.size()), mass_list.data(),
+               eff_value.data(), eff_error_x.data(), eff_error_x.data(),eff_errorLow.data(),
+               eff_errorUp.data());
+    }
+
+    static std::shared_ptr<TGraph> CreateRatio(const TGraphAsymmErrors& nume, const TGraphAsymmErrors& deno)
+    {
+        const size_t n_points = static_cast<size_t>(nume.GetN());
+        std::vector<double> x, y;
+        for(size_t n = 0; n < n_points; n++){
+            if(deno.GetY()[n] != 0) {
+                x.push_back(deno.GetX()[n]);
+                y.push_back(nume.GetY()[n] / deno.GetY()[n]);
+            }
+        }
+        return std::make_shared<TGraph>(x.size(), x.data(), y.data());
+    }
+
+    void PrintTGraph(const std::vector<std::shared_ptr<TGraphAsymmErrors>>& gr,
+                     const std::vector<std::shared_ptr<TGraph>>& ratioGraphs)
+    {
+        if(trigger_descs.size() == 1) {
+
+            canvas.cd();
```

Comment:
there are many parts in common between 1 and >1 cases => remove duplicated code.